packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -190,8 +197,33 @@ export class ChainService implements ChainServiceInterface {
       to: nitroAdjudicatorAddress,
     };
 
-    const captureExpectedErrors = (reason: any) => {
-      if (!reason.error.message.includes('revert Channel finalized')) throw reason;
+    const captureExpectedErrors = async (reason: any) => {
+      if (reason.error?.message.includes('Channel finalized')) {
+        return;
+      }
+      const firstState = finalizationProof[0];
+      const [
+        _turnNumRecord,
+        finalizesAt,
+        _fingerprint,
+      ] = await this.nitroAdjudicator.getChannelStorage(
+        getChannelId({
+          ...firstState,
+          participants: firstState.participants.map(p => p.signingAddress),
+        })
+      );
+
+      // Check if the channel has been finalized in the past or will finalize in the next 10 minutes
+      const timeNowSeconds = Date.now() / 1_000;
+      if (
+        finalizesAt &&
+        (timeNowSeconds - Number(finalizesAt) >= 0 ||
+          Number(finalizesAt) - timeNowSeconds <= 60 * 10)
```

Comment:
If the channel finalizes in the next 10 minutes, that indicates that there is a _challenge underway_. Do we really want to return in that case? Is that an expected behaviour in the happy path?

Is this instead due to the fact that `timeNowSeconds` is an approximation of `latestBlock.timestamp`?

Reply:
This case represents the following:
1. We have submitted a conclude and transfer all transaction and that transaction has failed
1. We have queried the chain and found out that the channel storage has a `finalizesAt` timestamp.

We need a heuristic to account for skew between timestamps onchain and timestamps offchain. In this PR the heuristic is to allow for skew of 10 minutes. Perhaps this is too large of an allowance?

If we want to consider challenges in this logic, we probably have to compare the fingerprint to the latest state in the wallet. This is not trivial as the chain service does not store states.

Thoughts?

Reply:
Actually, I see you comment below about not using a local timestamp. So I think we might be able to remove this skew allowance.

Reply:
There may still be some issues due to chain reorgs. Something like
1. I submit tx1.
2. It is reverted in block 10 because a transaction tx2 succeeded in block 9.
3. The chain is forked at block 9 and no longer includes tx1 or tx2.
4. I get the channel state from block 11 in this fork.

In this specific case, we would want to observe that the channel was not concluded and retry, so I think the overall logic in this pr is still correct. But I'm not sure I have a full grasp on the possible edge cases.

Reply:
I guess, to clarify, I think the approach in this pr is sufficient for now, and probably a good approach, but it should just use the block timestamp rather than the system's timestamp.

Reply:
9d1de309a.

Note that this PR does not attempt to retry transactions. The chain service throws an error if a transaction fails for any reason besides acceptable failure reasons. Transaction retries are something that would be nice to have down the line.